src/policy/marksweepspace/native_ms/block.rs

```diff
@@ -0,0 +1,450 @@
+                    cell,
+                    0,
+                    Ordering::SeqCst,
+                );
+                unsafe {
+                    cell.store::<Address>(last);
+                }
+                last = cell;
+            }
+            cell += cell_size;
+        }
+
+        self.store_free_list(last);
+    }
+
+    /// This is a naive implementation that is inefficient but should be correct.
+    /// In this implementation, we simply go through each possible object
+    /// reference and see if it has the mark bit set. If we find mark bit, that means the cell is alive. If we didn't find
+    /// the mark bit in the entire cell, it means the cell is dead.
+    fn naive_brute_force_sweep<VM: VMBinding>(&self) {
+        use crate::util::constants::MIN_OBJECT_SIZE;
+
+        // Cell size for this block.
+        let cell_size = self.load_block_cell_size();
+        // Current cell
+        let mut cell = self.start();
+        // Last free cell in the free list
+        let mut last = Address::ZERO;
+        // Current cursor
+        let mut cursor = cell;
+
+        debug!("Sweep block {:?}, cell size {}", self, cell_size);
+
+        while cell + cell_size <= self.end() {
+            // possible object ref
+            let potential_object_ref = ObjectReference::from_raw_address(
+                cursor + VM::VMObjectModel::OBJECT_REF_OFFSET_LOWER_BOUND,
+            );
+            trace!(
+                "{:?}: cell = {}, last cell in free list = {}, cursor = {}, potential object = {}",
+                self,
+                cell,
+                last,
+                cursor,
+                potential_object_ref
+            );
+
+            if VM::VMObjectModel::LOCAL_MARK_BIT_SPEC
+                .is_marked::<VM>(potential_object_ref, Ordering::SeqCst)
+            {
+                debug!("{:?} Live cell: {}", self, cell);
+                // If the mark bit is set, the cell is alive.
+                // We directly jump to the end of the cell.
+                cell += cell_size;
+                cursor = cell;
+            } else {
+                // If the mark bit is not set, we don't know if the cell is alive or not. We keep search for the mark bit.
+                cursor += MIN_OBJECT_SIZE;
+
+                if cursor >= cell + cell_size {
+                    // We now stepped to the next cell. This means we did not find mark bit in the current cell, and we can add this cell to free list.
+                    debug!(
+                        "{:?} Free cell: {}, last cell in freelist is {}",
+                        self, cell, last
+                    );
+
+                    // Clear alloc bit: we don't know where the object reference actually is, so we bulk zero the possible area.
+                    #[cfg(feature = "global_alloc_bit")]
+                    crate::util::alloc_bit::ALLOC_SIDE_METADATA_SPEC.bzero_metadata(
```

Comment:
The current implementation of `bzero_metadata` can only clear whole bytes in the metadata, but the VO-bit range of a cell may not fill whole bytes, and may span multiple bytes.  This will cause this `bzero_metadata` call to zero more or less bits than actually needed.

Reply:
And it is preferrable to call `crate::util::alloc_bit::bzero_alloc_bit` to ensure all accesses to the VO-bit metadata are encapsulated.

Reply:
Feels like this is a bug in `bzero_metadata`. I can fix it in a separate PR. Is there a way to work around the bug in this PR?

Reply:
One possible workaround is not using `bzero_metadata`. Instead, use `crate::util::alloc_bit::ALLOC_SIDE_METADATA_SPEC.set_zero` to clear one bit at a time.  The granularity of `ALLOC_SIDE_METADATA_SPEC` is one bit per word (8 bytes).  For the largest size class in mimalloc, it is 8192 bits for 8192 words.

Reply:
I submitted a PR to fix the issue: https://github.com/mmtk/mmtk-core/pull/707

Reply:
I also changed this to use `bzero_alloc_bit`.